src/circuits/joinsplit.tcc

```diff
@@ -41,11 +42,19 @@ class joinsplit_gadget : libsnark::gadget<FieldT> {
         std::array<std::shared_ptr<libsnark::digest_variable<FieldT> >, NumOutputs> output_commitments; // List of commitments generated for the new notes
         libsnark::pb_variable_array<FieldT> zk_vpub_in; // Public value that is put into the mix
         libsnark::pb_variable_array<FieldT> zk_vpub_out; // Value that is taken out of the mix
+        std::shared_ptr<libsnark::digest_variable<FieldT> > h_sig;    // sighash h Sig := hSigCRH(randomSeed, {nf_old} , joinSplitPubKey ) (p.53 ZCash proto. spec.)
+        std::array<std::shared_ptr<libsnark::digest_variable<FieldT> >, NumInputs> h_is; // List of malleability tags
 
         // ---- Auxiliary inputs (private) ---- //
         libsnark::pb_variable_array<FieldT> zk_total_uint64; // Total amount transfered in the transaction
-        std::array<std::shared_ptr<input_note_gadget<FieldT, HashT, HashTreeT>>, NumInputs> input_notes; // Input note gadgets
-        std::array<std::shared_ptr<output_note_gadget<FieldT, HashT>>, NumOutputs> output_notes; // Output note gadgets
+        std::array<std::shared_ptr<input_note_gadget<FieldT, HashT, HashTreeT> >, NumInputs> input_notes; // Input note gadgets
+        std::array<std::shared_ptr<output_note_gadget<FieldT, HashT> >, NumOutputs> output_notes; // Output note gadgets
+        std::array<std::shared_ptr<PRF_pk_gadget<FieldT, HashT>>, NumInputs> h_is_gadgets; // Input note gadgets
+        std::array<std::shared_ptr<PRF_rho_gadget<FieldT, HashT> >, NumOutputs> rho_is_gadgets; // Output note gadgets
```

Comment:
The comment does not match with the gadget used

Reply:
What about moving these gadgets in the output note gadgets, like here: https://github.com/zcash/zcash/blob/master/src/zcash/circuit/note.tcc#L174? This would require to pass `phi` and `h_sig` as arguments to the gadget constructor, to enable the computation of `rho` inside the output note gadget directly.

Reply:
I have changed the name and the ordering.

As for moving the `rho_i_gadgets` to inside the output note gadget, I did not do it because I did not want to order the output note (having `i` as argument). If you think it is better to compute it inside the gadget no problem.

Reply:
In that case for consistency we should also have the `h_i_gadgets` in the input note.

Reply:
> As for moving the rho_i_gadgets to inside the output note gadget, I did not do it because I did not want to order the output note (having i as argument)

Hum.. I'm not sure I understand what you mean. You are already following a certain order, right? Since you pass `rho_is[i]` as input to `output_notes[i]`, right?

Reply:
Yes you are right.

It was mostly nitpicking I guess. I just slightly disliked the idea of having an explicit "i" in the output note interface.

Reply:
Well we can keep it that way for now. It's not related to this PR anyway, but we can keep this in mind.
Just make sure that you grouped the gadgets as specified above please.

Reply:
Sure !

Reply:
Done :)